### PR TITLE
Refactored pandas_timedelta_to_timedeltastruct

### PR DESCRIPTION
**Pull request by WillAyd**
this mostly cuts down on copy paste, but also implements `const` where it can be done on some variables

### CODE REVIEW

File: pandas/_libs/src/vendored/numpy/datetime/np_datetime.c [hunk 1]
```diff
@@ -711,355 +711,120 @@ void pandas_datetime_to_datetimestruct(npy_datetime dt, NPY_DATETIMEUNIT base,
...
-      ifrac -= out->ms * 1000LL;
-      out->us = ifrac / 1L;
-      ifrac -= out->us * 1L;
-      out->ns = ifrac;
-    } else {
-      out->ms = 0;
-      out->us = 0;
-      out->ns = 0;
-    }
+  case (NPY_FR_D):
```
Comment (jbrockmendel) [inline pandas/_libs/src/vendored/numpy/datetime/np_datetime.c:724]: is there a reason why some cases are in parens and others are not?

File: pandas/_libs/src/vendored/numpy/datetime/np_datetime.c [hunk 2]
```diff
@@ -711,355 +711,120 @@ void pandas_datetime_to_datetimestruct(npy_datetime dt, NPY_DATETIMEUNIT base,
...
+  case NPY_FR_us:
+  case NPY_FR_ns: {
+    const npy_int64 sec_per_day = 86400;
+    npy_int64 per_sec;
+    if (base == NPY_FR_s) {
+      per_sec = 1;
+    } else if (base == NPY_FR_ms) {
+      per_sec = 1000;
+    } else if (base == NPY_FR_us) {
+      per_sec = 1000000;
```
Comment (jbrockmendel) [inline pandas/_libs/src/vendored/numpy/datetime/np_datetime.c:750]: overlap with tslibs.dtypes.periods_per_second.  probably not worth the hassle of de-duplicating

**Comment by jbrockmendel** [comment]
Small question, otherwise looks like a nice cleanup

File: pandas/_libs/src/vendored/numpy/datetime/np_datetime.c (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/src/vendored/numpy/datetime/np_datetime.c:724]: No - not sure how the parentheses snuck in here. Will remove

File: pandas/_libs/src/vendored/numpy/datetime/np_datetime.c (hunk 2 above)
Comment (WillAyd) [reply pandas/_libs/src/vendored/numpy/datetime/np_datetime.c:750]: If we wanted to we could make that an inline function like:

```c
inline npy_int64 per_sec_multiplier(NPY_DATETIMEUNIT base):
    switch base:
    case NPY_FR_s:
        return 1;
    case NPY_FR_ms:
        return 1000;
    ...
```

File: pandas/_libs/src/vendored/numpy/datetime/np_datetime.c (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/src/vendored/numpy/datetime/np_datetime.c:712]: LGTM once these are cleaned up (and green)

**Comment by WillAyd** [comment]
Sorry I let this slip - this should be reviewable now

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @WillAyd

**Pull request merged**